File: ts/src/Employee.ts
const assert = require('assert').strict

export class Employee {
  public workHours: number[]
  public name: string

  constructor(name: string, workHours: number[]) {
    assert.ok(
      workHours.filter((hour) => Number.isInteger(hour)).length ===
        workHours.length,
      new Error('工時資料應該為數字')
    )
    assert.ok(workHours.length === 7, new Error('應該要有一週的工時資料'))
    assert.ok(
      assertRange(0, 16)(workHours),
      new Error('工時資料應該為 0 - 16 的數字')
    )

    this.name = name
    this.workHours = workHours
  }

  // CFO
  public calculatePay(): number {
    return this.regularHours * 200 + this.overtimeHours * 350
  }

  // COO
  public reportHours(): string {
    return `Regular Hours: ${this.regularHours}`
  }

  // CTO
  public save() {
    return `saved`
  }

  private get regularHours(): number {
    return this.workHours.reduce((acc, hour, index) => {
      if (index === 0 || index === 6) {
        return acc
      }

      return acc + Math.min(hour, 8)
    }, 0)
  }

  private get overtimeHours(): number {
    const workdays = [1, 2, 3, 4, 5]
    return this.workHours.reduce((acc, hour, index) => {
      const isWorkDay = !!workdays.find((day) => day === index)
      if (isWorkDay) {
        return acc + Math.max(hour - 8, 0)
      } else {
        return acc + hour
      }
    }, 0)
  }
}

const assertRange = (start: number, end: number) => (arr: number[]) =>
  arr.filter((v) => v >= start && v <= end).length === arr.length

File: java/src/main/java/community/dddtw/refactor/Employee.java
package community.dddtw.refactor;

import java.util.ArrayList;
import java.util.Arrays;

public class Employee {
    public int[] workHours;
    public String name;

    public Employee(String name, int[] workHours) {
        if (workHours.length != 7) {
            throw new IllegalArgumentException("應該要有一週的工時資料");
        }
        for (int workHour : workHours) {
            if (workHour < 0 || workHour > 16) {
                throw new IllegalArgumentException("工時資料應該為 0 - 16 的數字");
            }
        }

        this.workHours = workHours;
        this.name = name;
    }


    // CFO
    public int calculatePay() {
        return this.regularHours() * 200 + this.overtimeHours() * 350;
    }

    // COO
    public String reportHours() {
        return "Regular Hours: " + this.regularHours();
    }

    // CTO
    public String save() {
        return "saved";
    }


    private int overtimeHours() {
        int hourCounter = 0;
        ArrayList<Integer> workDays = new ArrayList<Integer>(Arrays.asList(1, 2, 3, 4, 5));
        for (int i = 0; i < this.workHours.length; i++) {
            int workDay = i;
            int workHour = this.workHours[i];
            if (workDays.contains(workDay)) {
                hourCounter += Math.max(workHour - 8, 0);
            } else {
                hourCounter += workHour;
            }
        }
        return hourCounter;
    }

    private int regularHours() {
        int hourCounter = 0;
        for (int i = 0; i < this.workHours.length; i++) {
            int workHour = this.workHours[i];
            if (i == 0 || i == 6) {
                continue;
            }
            hourCounter += Math.min(workHour, 8);
        }
        return hourCounter;
    }
}


